server/EXCEPTIONS.md

```diff
@@ -0,0 +1,192 @@
+# Exception handling in our backend
+
+## Quickstart
+
+To generate errors in the backend that will be passed to the
+frontend, throw a [Nest-defined HttpException](https://docs.nestjs.com/exception-filters#throwing-standard-exceptions), <u>_generally within a service_</u>:
+
+```js
+  throw new HttpException(responseBody, status)
+```
+
+Where...
+  - `status` is one of `HttpStatus` enum codes. E.g. `HttpStatus.UNAUTHORIZED`
+  - `responseBody` argument is an object with the following schema.
+    ```js
+       const responseBody = {
+        code: 'APPLICANT_PORTAL_ERROR_CODE',
+        message: 'Applicant Portal could not find coffee. CRM: CRM ran out of tea.',
+      }
+    ```
+    - The `code` property is developer-defined, and should be a unique, concise,
+    descriptive identifier for the type of error. Users can provide this code to
+    help developers locate the code that yielded the error.
+    - The `message` property is a short, human-readable sentence firstly
+    describing the error for the benefit of end users. Secondly, it should contain
+    errors relayed from CRM, if available.
```

Comment:
Sorry, just now seeing this, but this isn't an allowed key name in the JSON:API error response: https://jsonapi.org/format/#errors

Please be sure to check existing standards before extending them. If we need our own key names, JSON:API allows for a `meta` property whose value is an object with whatever non-standard key names we need. In this case, I think the spec provides a nice option for this, namely in `detail`.

> detail: a human-readable explanation specific to this occurrence of the problem. Like title, this field’s value can be localized.

Reply:
@allthesignals  the top-level `message` property isn't something I introduced. It's something that already exists, not sure what part of the code is generating it.

Reply:
Can you point to where it's being generated? Not sure I know where to find that... our Http Exception filter itself just wraps the error in an array.

Reply:
Okay, so I found out that it's generated automatically from the Nest HttpException class, but on the docs it's marked as _private_:

http://nestjs-doc.exceptionfound.com/classes/httpexception.html#error
That'll teach me to only rely on browsing response formats in the dev console

Reply:
@allthesignals There are two catches: 
  1. it seems Nest HTTPExceptions out of the box do not conform to JSONAPI Errors specification. According to the docs, by default when constructing a Nest HTTPException without overriding the response body (simply passing a string instead of an object to the first argument)...
```ts
throw new HttpException('Forbidden', HttpStatus.FORBIDDEN); 
```
then the generated error object looks like this:
```json
{
  "statusCode": 403,
  "message": "Forbidden"
}
```

Note that `message` and `statusCode` were both top-level properties automatically generated by Nest.

Strangely, when I try to verify this, the response looks slightly different in that `statusCode` is instead `status`:
```
     const errorMessage = `Error finding projects by contact ID. ${e.message}`;
     throw new HttpException(errorMessage, HttpStatus.UNAUTHORIZED);
```
... generates the following response:

```json
  {
    error: "Error finding projects by contact ID. Could not find a property named 'dcp_dcp_project_dcp_packae_project' on type 'Microsoft.Dynamics.CRM.dcp_project'.
 at ProjectsService.findManyByContactId (/Users/[...]/projects.service.js:49:19)",
    response: "Error finding projects by contact ID. Could not find a property named \'dcp_dcp_project_dcp_packae_project\' on type \'Microsoft.Dynamics.CRM.dcp_project\'.'",
    status: 401,
    message: "Error finding projects by contact ID. Could not find a property named \'dcp_dcp_project_dcp_packae_project\' on type \'Microsoft.Dynamics.CRM.dcp_project\'.'"
  }
```

Note that both in the docs, and in the live response, the `message` property was provided automatically by Nest. 
The live response does provide the `status` property, which conforms to the JSON API. 

2. Nest's preferred way of editing and adding error properties is to provide an object to the first argument:
```ts
@Get()
async findAll() {
  throw new HttpException({
    status: HttpStatus.FORBIDDEN,
    error: 'This is a custom message',
  }, HttpStatus.FORBIDDEN);
}
```
And the docs suggest this will place those properties in the top-level of the error:
```json
{
  "status": 403,
  "error": "This is a custom message"
}
```

However, in a live test, they are actually placed within a `response` property.
```ts
const errorMessage = `Error finding project by project ID. ${e.message}`;
      console.log(errorMessage);
      throw new HttpException({
        "code": "GET_PROJECT_SERVICE_ERROR",
        "detail": errorMessage,
        "title": "My title",
      }, HttpStatus.UNAUTHORIZED);
```
 yields
```
   {
    Error: Http Exception
      at ProjectsService.getProject (/Users/[...]/projects.service.js:78:19),
     response:
        { 
          code: 'GET_PROJECT_SERVICE_ERROR',
          detail: 'Error finding project by project ID. Could not find a property named \'dcp_dcp_project_dcp_pakage_project\' on type \'Microsoft.Dynamics.CRM.dcp_project\'.',
          title: 'My title'
        },
        status: 401,
        message: 'Http Exception'
   }
```


One thing we could do is modify our filter to mold HTTPException to comply to JSONAPI spec.

Perhaps:
```ts
@Catch(HttpException)
export class HttpExceptionFilter implements ExceptionFilter {
  catch(exception: HttpException, host: ArgumentsHost) {
    ...
    // this shapes the error responses into JSON:API
    // see https://jsonapi.org/format/#errors
    response
      .status(status)
      .json({
        errors: [{
            detail: exception.message,
            code: exception.response.status,
            status: exception.status,
         }],
      });
  }
}
```

Reply:
Oh! Interesting. I think that's exactly what has to happen...

@godfreyyeung thanks for sleuthing this... I didn't know Nest was opinionated about the shape of the exception. I guess it makes sense given that it's called an _HTTPException_, so some of the language is there.

I think re-mapping is the right way to go... but we might need to figure out how we want to handle each "feature" in the JSON:API error spec — what is the difference between a "title" and a "detail"? Should we use "meta"? Should we use "source"? And so on.

Anyway, great find!

Reply:
Yes, will revise the guide to account for this! 
Thinking about using "status", "detail" and "code" at least. 
Maybe also "meta" to hold controller/service information.